geth/node/manager.go

```diff
@@ -97,6 +94,12 @@ func (m *NodeManager) startNode(config *params.NodeConfig) (<-chan struct{}, err
 		m.node = ethNode
 		m.nodeStopped = make(chan struct{}, 1)
 		m.config = config
+
+		// init RPC client for this node
+		m.rpcClient, err = rpc.NewClient(m.node, m.config.UpstreamConfig.URL)
+		if err != nil {
+			log.Error("Init RPC client failed", "error", err)
```

Comment:
Or at least this way:
1. Return error in case of bad connection to the local node.
2. Log error in case of a failed connection to the upstream.
3. Return error when RPC request doesn't reach the upstream node.

Reply:
Hmm, I was assuming go-ethereum client will reconnect automatically, but that's true only after connection is established.

So what we should in case 2? Start some reconnection/retries probably?

Reply:
I'd just logged this and continued. The important thing is to have the connection when we are actually about to send a request. If it fails, let's report an error to the caller.

Reply:
Make sense. I will add tests to make sure this logic works as expected and underlying client is doing reconnects.

Reply:
@tiabc just added this logic. Take a look, please, I hope I didn't overcomplicate it.